File: apps/backend/app/schemas/scout.py
"""Contracts for scheduled job discovery and application-pack preparation."""

from typing import Any, Literal

from pydantic import BaseModel, Field, HttpUrl


class SearchProfileConfig(BaseModel):
    desired_titles: list[str] = Field(default_factory=list)
    adjacent_titles: list[str] = Field(default_factory=list)
    locations: list[str] = Field(default_factory=list)
    workplace_types: list[Literal["remote", "hybrid", "onsite"]] = Field(
        default_factory=list
    )
    seniority: list[str] = Field(default_factory=list)
    minimum_years_experience: int | None = Field(default=None, ge=0, le=60)
    maximum_years_experience: int | None = Field(default=None, ge=0, le=60)
    required_skills: list[str] = Field(default_factory=list)
    preferred_skills: list[str] = Field(default_factory=list)
    excluded_terms: list[str] = Field(default_factory=list)
    excluded_skills: list[str] = Field(default_factory=list)
    preferred_companies: list[str] = Field(default_factory=list)
    blocked_companies: list[str] = Field(default_factory=list)
    industries: list[str] = Field(default_factory=list)
    employment_types: list[str] = Field(default_factory=list)
    salary_floor: int | None = Field(default=None, ge=0)
    currency: str = "USD"
    visa_sponsorship_required: bool | None = None
    work_authorized_countries: list[str] = Field(default_factory=list)
    daily_pack_limit: int = Field(default=20, ge=1, le=20)
    reserve_limit: int = Field(default=10, ge=0, le=50)
    minimum_match_score: float = Field(default=70, ge=0, le=100)
    max_per_company: int = Field(default=2, ge=1, le=20)
    timezone: str = "Asia/Singapore"


class SearchProfileCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    config: SearchProfileConfig = Field(default_factory=SearchProfileConfig)
    is_active: bool = True


class SearchProfileUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    config: SearchProfileConfig | None = None
    is_active: bool | None = None


class SearchProfileResponse(SearchProfileCreate):
    profile_id: str
    user_id: str
    created_at: str
    updated_at: str


SourceType = Literal[
    "manual",
    "greenhouse",
    "lever",
    "ashby",
    "smartrecruiters",
    "careers_page",
    "email_alert",
    "linkedin_partner",
    "indeed_partner",
]


class JobSourceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    source_type: SourceType
    config: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class JobSourceResponse(JobSourceCreate):
    source_id: str
    user_id: str
    status: str
    last_scan_at: str | None = None
    last_error: str | None = None
    created_at: str
    updated_at: str


class ManualJobCreate(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    company: str = Field(min_length=1, max_length=300)
    description: str = Field(min_length=50)
    source_url: HttpUrl | None = None
    location: str | None = None
    workplace_type: str | None = None
    employment_type: str | None = None
    compensation: str | None = None
    profile_id: str | None = None


class EmailAlertIngest(BaseModel):
    subject: str
    text: str
    sender: str
    html: str | None = None


class JobPostingResponse(BaseModel):
    posting_id: str
    source_id: str | None = None
    source_type: str
    source_url: str | None = None
    apply_url: str | None = None
    company: str
    title: str
    location: str | None = None
    workplace_type: str | None = None
    employment_type: str | None = None
    compensation: str | None = None
    description: str
    status: str
    posted_at: str | None = None
    discovered_at: str


class JobMatchResponse(BaseModel):
    match_id: str
    profile_id: str
    posting: JobPostingResponse
    eligible: bool
    score: float
    score_components: dict[str, Any]
    gaps: list[Any]
    explanation: str | None = None
    state: str
    selected_date: str | None = None
    daily_rank: int | None = None
    pack_status: str | None = None
    referral_count: int = 0


class ArtifactPackResponse(BaseModel):
    pack_id: str
    match_id: str
    resume_id: str | None = None
    cover_letter: str | None = None
    outreach_message: str | None = None
    referral_subject: str | None = None
    referral_email: str | None = None
    linkedin_message: str | None = None
    interview_prep: dict[str, Any] | None = None
    application_answers: dict[str, Any] = Field(default_factory=dict)
    provenance: dict[str, Any] = Field(default_factory=dict)
    llm_usage: dict[str, Any] = Field(default_factory=dict)
    status: str
    error: str | None = None
    created_at: str
    updated_at: str


class ArtifactPackUpdate(BaseModel):
    cover_letter: str | None = Field(default=None, max_length=30_000)
    outreach_message: str | None = Field(default=None, max_length=10_000)
    referral_subject: str | None = Field(default=None, max_length=300)
    referral_email: str | None = Field(default=None, max_length=10_000)
    linkedin_message: str | None = Field(default=None, max_length=3000)


class PreparePackRequest(BaseModel):
    force: bool = False
    regeneration_key: str | None = Field(default=None, min_length=8, max_length=120)


class ContactResponse(BaseModel):
    contact_id: str
    first_name: str
    last_name: str
    email: str | None = None
    company: str | None = None
    position: str | None = None
    profile_url: str | None = None
    connected_on: str | None = None
    relationship_strength: int = 0
    notes: str | None = None
    excluded: bool = False


class ReferralMatchResponse(BaseModel):
    referral_match_id: str
    contact: ContactResponse
    score: float
    explanation: str
    contacted_at: str | None = None
    response_notes: str | None = None


class ReferralHistoryUpdate(BaseModel):
    contacted: bool | None = None
    response_notes: str | None = Field(default=None, max_length=4000)


class CandidateFactCreate(BaseModel):
    fact_key: str = Field(pattern=r"^[a-z0-9_]+$", min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=120)
    value: str = Field(min_length=1, max_length=4000)
    category: str = Field(default="general", max_length=80)
    sensitive: bool = False


class CandidateFactResponse(CandidateFactCreate):
    fact_id: str
    created_at: str
    updated_at: str


class ScoutRunResponse(BaseModel):
    run_id: str
    source_id: str
    status: str
    discovered_count: int
    created_count: int
    error: str | None = None
    started_at: str
    finished_at: str | None = None


class ContactImportResponse(BaseModel):
    imported: int
    updated: int
    skipped: int


class OpportunityStateUpdate(BaseModel):
    state: Literal[
        "discovered", "selected", "reserve", "preparing", "ready", "ignored", "applied"
    ]

File: apps/frontend/lib/api/scout.ts
import { apiDelete, apiFetch, apiPatch, apiPost } from './client';

export interface SearchProfileConfig {
  desired_titles?: string[];
  adjacent_titles?: string[];
  locations?: string[];
  workplace_types?: Array<'remote' | 'hybrid' | 'onsite'>;
  seniority?: string[];
  minimum_years_experience?: number | null;
  maximum_years_experience?: number | null;
  required_skills?: string[];
  preferred_skills?: string[];
  excluded_terms?: string[];
  preferred_companies?: string[];
  industries?: string[];
  employment_types?: string[];
  daily_pack_limit: number;
  reserve_limit?: number;
  minimum_match_score: number;
  max_per_company?: number;
  timezone?: string;
  [key: string]: unknown;
}

export interface SearchProfile {
  profile_id: string;
  name: string;
  config: SearchProfileConfig;
  is_active: boolean;
}

export interface JobPosting {
  posting_id: string;
  title: string;
  company: string;
  location: string | null;
  workplace_type: string | null;
  description: string;
  source_url: string | null;
  source_type: string;
  published_at: string | null;
}

export interface Opportunity {
  match_id: string;
  profile_id: string;
  posting: JobPosting;
  eligible: boolean;
  score: number;
  score_components: Record<string, number>;
  gaps: string[];
  explanation: string;
  state: string;
  selected_date: string | null;
  daily_rank: number | null;
  pack_status: string | null;
  referral_count: number;
}

export interface ArtifactPack {
  pack_id: string;
  match_id: string;
  resume_id: string | null;
  cover_letter: string | null;
  outreach_message: string | null;
  referral_subject: string | null;
  referral_email: string | null;
  linkedin_message: string | null;
  interview_prep: Record<string, unknown> | null;
  application_answers: Record<string, string>;
  provenance: Record<string, unknown>;
  llm_usage: {
    calls?: number;
    prompt_tokens?: number;
    completion_tokens?: number;
    cost_usd?: number;
  };
  status: string;
  error: string | null;
}

export interface JobSource {
  source_id: string;
  name: string;
  source_type: string;
  config: Record<string, unknown>;
  enabled: boolean;
  status: string;
  last_scan_at: string | null;
  last_error: string | null;
}

export interface ScoutRun {
  run_id: string;
  source_id: string;
  status: string;
  discovered_count: number;
  created_count: number;
  error: string | null;
  started_at: string;
  finished_at: string | null;
}

export interface CandidateFact {
  fact_id: string;
  fact_key: string;
  label: string;
  value: string;
  category: string;
  sensitive: boolean;
}

export interface Contact {
  contact_id: string;
  first_name: string;
  last_name: string;
  company: string | null;
  position: string | null;
  profile_url: string | null;
  email: string | null;
}

export interface ReferralMatch {
  referral_match_id: string;
  contact: Contact;
  score: number;
  explanation: string;
  contacted_at: string | null;
  response_notes: string | null;
}

async function json<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const payload = await response.json().catch(() => ({}));
    throw new Error(
      typeof payload.detail === 'string' ? payload.detail : `Request failed (${response.status})`
    );
  }
  return response.json() as Promise<T>;
}

export const listProfiles = () => apiFetch('/search-profiles').then(json<SearchProfile[]>);
export const createProfile = (payload: Record<string, unknown>) =>
  apiPost('/search-profiles', payload).then(json<SearchProfile>);
export const updateProfile = (id: string, payload: Record<string, unknown>) =>
  apiPatch(`/search-profiles/${id}`, payload).then(json<SearchProfile>);
export const listOpportunities = () => apiFetch('/opportunities').then(json<Opportunity[]>);
export const listToday = () => apiFetch('/opportunities/today').then(json<Opportunity[]>);
export const getPack = (id: string) =>
  apiFetch(`/opportunities/${id}/pack`).then(json<ArtifactPack>);
export const updatePack = (id: string, payload: Record<string, string>) =>
  apiPatch(`/opportunities/${id}/pack`, payload).then(json<ArtifactPack>);
export const preparePack = (id: string, regenerate = false) =>
  apiPost(`/opportunities/${id}/prepare`, {
    force: regenerate,
    regeneration_key: regenerate ? crypto.randomUUID() : null,
  }).then(json<ArtifactPack>);
export const updateOpportunity = (id: string, state: string) =>
  apiPatch(`/opportunities/${id}`, { state }).then(json<Record<string, unknown>>);
export const replaceOpportunity = (id: string) =>
  apiPost(`/opportunities/${id}/replace`, {}).then(json<Record<string, unknown>>);
export const selectToday = (profileId: string) =>
  apiPost(`/opportunities/today/select?profile_id=${encodeURIComponent(profileId)}`, {}).then(
    json<Opportunity[]>
  );
export const addManualJob = (payload: Record<string, unknown>) =>
  apiPost('/opportunities/manual', payload).then(json<{ posting_id: string }>);
export const listSources = () => apiFetch('/sources').then(json<JobSource[]>);
export const listScoutRuns = () => apiFetch('/scout-runs').then(json<ScoutRun[]>);
export const createSource = (payload: Record<string, unknown>) =>
  apiPost('/sources', payload).then(json<JobSource>);
export const scanSource = (id: string) =>
  apiPost(`/sources/${id}/scan`, {}).then(json<Record<string, unknown>>);
export const listContacts = () => apiFetch('/contacts').then(json<Contact[]>);
export const listCandidateFacts = () => apiFetch('/candidate-facts').then(json<CandidateFact[]>);
export const saveCandidateFact = (payload: Record<string, unknown>) =>
  apiPost('/candidate-facts', payload).then(json<CandidateFact>);
export const importMasterResumeFacts = () =>
  apiPost('/candidate-facts/import-master', {}).then(
    json<{ imported: number; facts: CandidateFact[] }>
  );
export const deleteCandidateFact = (id: string) =>
  apiDelete(`/candidate-facts/${id}`).then(json<{ message: string }>);
export const listReferrals = (id: string, refresh = false) =>
  apiFetch(`/opportunities/${id}/referrals?refresh=${refresh}`).then(json<ReferralMatch[]>);
export const updateReferral = (
  matchId: string,
  referralId: string,
  payload: { contacted?: boolean; response_notes?: string }
) =>
  apiPatch(`/opportunities/${matchId}/referrals/${referralId}`, payload).then(json<ReferralMatch>);
export async function importConnections(
  file: File
): Promise<{ imported: number; updated: number }> {
  const form = new FormData();
  form.append('file', file);
  return apiFetch('/contacts/import/linkedin', { method: 'POST', body: form }).then(
    json<{ imported: number; updated: number }>
  );
}

export async function streamScoutEvents(onEvent: () => void, signal: AbortSignal): Promise<void> {
  const response = await apiFetch('/events', { signal }, 1_800_000);
  if (!response.ok || !response.body) throw new Error('Live progress is unavailable.');
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  while (!signal.aborted) {
    const { done, value } = await reader.read();
    if (done) return;
    buffer += decoder.decode(value, { stream: true });
    const messages = buffer.split('\n\n');
    buffer = messages.pop() || '';
    for (const message of messages) {
      if (message.includes('event: scout')) onEvent();
    }
  }
}
